### PR TITLE
Adding MS SQL translation

### PR DESCRIPTION
**Pull request by edgararuiz**
New translation allows dplyr to select, group, summarise and join

### CODE REVIEW

**Comment by imanuelcostigan** [comment]
@edgararuiz have you looked at [RSQLServer](https://github.com/imanuelcostigan/RSQLServer)? I incorporated a dplyr backend in this package because dplyr wasn't going to include new backends in it

**Comment by edgararuiz** [comment]
Hi @imanuelcostigan, thank you.  This translation is primarily meant to work with the 
[odbc](https://github.com/rstats-db/odbc) package

File: R/db-mssql.R [hunk 1]
```diff
@@ -0,0 +1,152 @@
+#' @export
+`db_desc.Microsoft SQL Server` <- function(x) {
+  info <- DBI::dbGetInfo(x)
+  host <- if (info$host == "") "localhost" else info$host
+
+  paste0("Microsoft SQL Server ", info$serverVersion, " [", info$user, "@",
```
Comment (hadley) [inline R/db-mssql.R]: Can you please follow the tidyverse indenting rules, i.e. http://style.tidyverse.org/syntax.html#long-lines ?

File: R/db-mssql.R [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                                             ...) {
+  out <- vector("list", 7)
+  names(out) <- c("select", "from", "where", "group_by",
+                  "having", "order_by","limit")
+
+  assert_that(is.character(select), length(select) > 0L)
+  out$select <- build_sql(
+    "SELECT ",
+
+    if (!is.null(limit) && !identical(limit, Inf)) {
```
Comment (hadley) [inline R/db-mssql.R]: It would be useful to have a comment here describing where SQL server deviates from SQL-92

File: R/db-mssql.R [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+      escape(order_by, collapse = ", ", con = con)
+    )
+  }
+
+  escape(unname(compact(out)), collapse = "\n", parens = FALSE, con = con)
+}
+
+
+#' @export
+`sql_subquery.Microsoft SQL Server` <- function(con, from, name = dbplyr:::unique_name(), ...) {
```
Comment (hadley) [inline R/db-mssql.R]: Isn't this the same as `sql_subquery.DBIConnection`?  You don't need to override methods that are the same as for DBIConnection.

File: R/db-mssql.R [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+`sql_translate_env.Microsoft SQL Server` <- function(con) {
+  dbplyr::sql_variant(
+    scalar = dbplyr::sql_translator(
+      .parent = dbplyr::base_scalar,
+      as.numeric = function(x) build_sql("CAST(", x, " AS NUMERIC)"),
+      as.double  = function(x) build_sql("CAST(", x, " AS NUMERIC)"),
+      as.integer  = function(x) build_sql("CAST(", x, " AS INT)"),
+      as.logical = function(x) build_sql("CAST(", x, " AS BOOLEAN)"),
+      as.character  = function(x) build_sql("CAST(", x, " AS VARCHAR(MAX))"),
+      as.date  = function(x) build_sql("CAST(", x, " AS DATE)"),
```
Comment (hadley) [inline R/db-mssql.R]: There's no `as.date()` in base R?

File: R/db-mssql.R [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+      as.Date  = function(x) build_sql("CAST(", x, " AS DATE)"),
+      paste = function(..., sep = " ") build_sql("CONCAT_WS", list(sep, ...)),
+      paste0 = function(...) build_sql("CONCAT", list(...)),
+      xor = function(x, y) build_sql(x, " ^ ", y),
+      or = function(x, y) build_sql(x, " or ", y),
+      and = function(x, y) build_sql(x, " and ", y),
+      pmin = function(...) build_sql_if_compare(..., con = con, compare = "<="),
```
Comment (hadley) [inline R/db-mssql.R]: Did you forget to check in this function?

File: R/db-mssql.R [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+      n_distinct = function(...) dbplyr::build_sql("COUNT(DISTINCT", list(...), ")"),
+      sd =  function(...) dbplyr::build_sql("STDEV(", list(...), ")"),
+      var = function(...) dbplyr::build_sql("VAR(", list(...), ")")
+    )
+  )}
+
+
+
+#' @export
+`sql_join.Microsoft SQL Server` <- function(con, x, y, vars, type = "inner", by = NULL, ...) {
```
Comment (hadley) [inline R/db-mssql.R]: I think you can delete this too

**Comment by edgararuiz** [comment]
Still working on the simulate_mssql tests,  not sure if they're required to pass this PR.

File: R/simulate.r [hunk 7]
```diff
@@ -32,3 +32,26 @@ simulate_postgres <- function() {
     class = c("PostgreSQLConnection", "DBIConnection")
   )
 }
+
+# odbc connections --------------------------------------------------------
+
+setClass(
```
Comment (hadley) [inline R/simulate.r]: Can you please just simulate like the functions above?

File: R/db-mssql.R [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+      # MSSQL supports CONCAT_WS in the CTP version of 2016
+      .parent = dbplyr::base_scalar,
+      as.numeric = function(x) build_sql("CAST(", x, " AS NUMERIC)"),
+      as.double  = function(x) build_sql("CAST(", x, " AS NUMERIC)"),
+      as.integer  = function(x) build_sql("CAST(", x, " AS INT)"),
+      as.logical = function(x) build_sql("CAST(", x, " AS BOOLEAN)"),
+      as.character  = function(x) build_sql("CAST(", x, " AS VARCHAR(MAX))"),
+      as.Date  = function(x) build_sql("CAST(", x, " AS DATE)"),
+      paste0 = function(...) build_sql("CONCAT", list(...)),
+      or = function(x, y) build_sql(x, " or ", y),
```
Comment (hadley) [inline R/db-mssql.R]: Aren't these the default translations?

File: tests/testthat/test-mssql.r [hunk 9]
```diff
@@ -0,0 +1,46 @@
+context("MSSQL translation")
+
+# Initialize simulated DB connection -------------------------------------
+
+mssql_connection <- dbplyr:::simulate_odbc("Microsoft SQL Server")
```
Comment (hadley) [inline tests/testthat/test-mssql.r]: Can you please copy and paste into the individual tests? It's a bit hacky but it makes each test easier to explore interactively in isolation.

File: R/simulate.r (hunk 7 above)
Comment (edgararuiz) [inline R/simulate.r]: Sure, quick question, do want me to use a single use function called `simulate_mssql`, or use the same code pattern as the other connections?:
```
simulate_postgres <- function() {
  structure(
    list(),
    class = c("PostgreSQLConnection", "DBIConnection")
  )
}
```

Comment (hadley) [inline R/simulate.r]: New function - you'll need to mimic the classes that odbc generates

Comment (edgararuiz) [inline R/simulate.r]: Got it, thanks

File: R/db-mssql.R (hunk 8 above)
Comment (edgararuiz) [inline R/db-mssql.R]: OR and AND were the only ones that passed the tests w/o translation, I removed them

**Comment by edgararuiz** [comment]
Hi @hadley , the latest commit passed two of the three Travis CI jobs.  It looks like the previously merged PR (#7) had the same issue, so I don't think it's due to something that needs to be fixed in this PR.  The error message mentions a mismatched dplyr versions inside the job's environment.  

Regarding the updates, I removed all of the code I previously used to simulate a MS SQL connection, and am using your new tbl_lazy method.  I also added a new test to confirm that `head()` translates to TOP

File: tests/testthat/test-mssql.r [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+                    sql("VAR((`field_name`))"))
+})
+
+
+test_that("query uses TOP instead of LIMIT ", {
+  df <- data.frame(x = 1, y = 2)
+  df_mssql <- tbl_lazy(df, src = simulate_mssql())
+  expect_equivalent(show_query(head(df_mssql)),
+                    sql("SELECT  TOP 6 *
+FROM `df`"))
```
Comment (hadley) [inline tests/testthat/test-mssql.r]: Can you please use `\n` instead of the literal line break?

File: tests/testthat/test-mssql.r [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+                    sql("COUNT(DISTINCT(`field_name`))"))
+})
+test_that("sd aggregate functions translate to MSSQL STDEV ", {
+  expect_equivalent(translate_sql(sd(field_name),
+                                  window = FALSE,
+                                  con = mssql_connection),
+                    sql("STDEV((`field_name`))"))
+})
+test_that("var aggregate functions translate to MSSQL VAR ", {
+  expect_equivalent(translate_sql(var(field_name),
```
Comment (hadley) [inline tests/testthat/test-mssql.r]: Can you please use tidyverse style indenting? http://style.tidyverse.org/syntax.html#long-lines

File: NEWS.md [hunk 12]
```diff
@@ -150,6 +150,8 @@
   `win_current_group()` and `win_current_order()` are now exported. This
   should make it easier to provide customised SQL for window functions
   (#2051, #2126).
+  
+*  SQL translation for Microsoft SQL Server
```
Comment (hadley) [inline NEWS.md]: Can you please add `(@edgaruiz)`?

**Comment by hadley** [comment]
Thanks!

**Comment by edgararuiz** [comment]
Thank you for all the guidance!

**Pull request merged**